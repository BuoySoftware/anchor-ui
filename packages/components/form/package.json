{
  "name": "@buoysoftware/anchor-form",
  "version": "0.14.0-alpha.0",
  "main": "src/index.ts",
  "license": "MIT",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsc -outDir dist",
    "compile": "tsc --noEmit",
    "test": "jest",
    "postpack": "clean-package restore",
    "prepack": "clean-package"
  },
  "dependencies": {
    "@buoysoftware/anchor-layout": "^0.14.0-alpha.0",
    "@buoysoftware/anchor-typography": "^0.14.0-alpha.0",
    "i18next": "^22.4.9",
    "lodash": "*",
    "react-hook-form": "^7.42.1",
    "react-i18next": "^12.1.4"
  },
  "devDependencies": {
    "jest": "^29.3.1",
    "ts-jest": "^29.0.4",
    "ts-node": "^10.9.1"
  }
}
