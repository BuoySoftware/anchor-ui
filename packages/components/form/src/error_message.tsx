import { WarningCircle } from "iconoir-react";
import { Box, BoxProps, Flex } from "@buoysoftware/anchor-layout";
import { Body, Markdown } from "@buoysoftware/anchor-typography";
import { useTheme } from "@buoysoftware/anchor-theme";

interface OwnProps {
  errors: string[];
  parentId: string;
}

type ErrorMessageProps = OwnProps & BoxProps;

export const ErrorMessage: React.FC<ErrorMessageProps> = ({
  errors,
  parentId,
  ...props
}): React.ReactElement => {
  const { colors } = useTheme();
  const id = `error-${parentId}`;

  return (
    <Flex
      bg="background.criticalSubdued"
      borderRadius="4px"
      data-testid={id}
      id={id}
      py="xxs"
      px="xs"
      role="alert"
      {...props}
    >
      <WarningCircle color={colors.text.critical} width={16} height={16} />
      <Box ml="xs">
        {errors.map((error, index) => (
          <Body as="div" size="m" key={`error-${index}`} color="text.critical">
            <Markdown>{error}</Markdown>
          </Body>
        ))}
      </Box>
    </Flex>
  );
};
