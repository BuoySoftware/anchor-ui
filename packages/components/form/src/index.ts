export * from "./checkbox";
export * from "./checkbox_field";
export * from "./error_message";
export * from "./form";
export * from "./form_base_errors";
export * from "./form_button";
export * from "./form_scope_provider";
export * from "./input";
export * from "./label";
export * from "./radio_field";
export * from "./select";
export * from "./select_field";
export * from "./text_area";
export * from "./text_field";
export * from "./use_conditional_field";
export * from "./use_form_field";

export { default as anchorFormsEn } from "./i18n/anchorForms.en.json";
