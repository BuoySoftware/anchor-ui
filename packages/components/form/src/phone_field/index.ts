export * from "./phone_field";
