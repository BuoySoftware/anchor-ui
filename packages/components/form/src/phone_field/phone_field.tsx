import {
  isPossiblePhoneNumber,
  parsePhoneNumberFromString,
} from "libphonenumber-js";
import { useFormContext, Controller, RegisterOptions } from "react-hook-form";
import snakeCase from "lodash/snakeCase";
import { useTranslation } from "react-i18next";

import { FormField } from "../form_field";
import { useFormField } from "../use_form_field";
import PhoneInput, { PhoneInputProps } from "./phone_input";

interface OwnProps {
  defaultValue?: string;
  label?: Maybe<string>;
  name: string;
  rules?: Exclude<RegisterOptions, "valueAsNumber" | "valueAsDate">;
}

export type PhoneFieldProps = OwnProps &
  Omit<PhoneInputProps, "onChange" | "defaultValue">;

const formattedValue = (value: string): string => {
  const parsedPhoneNumber = parsePhoneNumberFromString(value);
  return parsedPhoneNumber?.formatNational() ?? "";
};

export const PhoneField: React.FC<PhoneFieldProps> = ({
  defaultValue = "",
  label: labelProp,
  name,
  rules,
  width = "input",
  zIndex = "formField",
  ...props
}): React.ReactElement => {
  const { control } = useFormContext();
  const { t } = useTranslation(["anchorForms", "forms"]);
  const {
    error,
    hint,
    label: formFieldLabel,
    placeholder,
  } = useFormField({
    defaultPlaceholder:
      t([
        "anchorForms:placeholders.phone_number",
        "forms:placeholders.phone_number",
      ]) ?? "",
    name,
  });
  const label = labelProp ?? formFieldLabel;
  const inputId = snakeCase(name);

  const phoneNumberValidation = (phoneNumber: string): string | undefined => {
    const valid = isPossiblePhoneNumber(phoneNumber, "US");

    if (!valid) {
      return (
        t("errors.invalidPhoneNumber", { ns: ["anchorForms", "forms"] }) ||
        undefined
      );
    }
  };

  return (
    <FormField
      data-testid={`phone-field-${inputId}`}
      error={error}
      hint={hint}
      inputId={inputId}
      inputType="phone"
      label={label}
      name={name}
      position="relative"
      zIndex={zIndex}
      width={width}
    >
      <Controller
        control={control}
        name={name}
        defaultValue={formattedValue(defaultValue)}
        render={({ field }) => {
          return (
            <PhoneInput
              id={inputId}
              placeholder={placeholder}
              {...props}
              {...field}
            />
          );
        }}
        rules={{ ...rules, validate: { phoneNumber: phoneNumberValidation } }}
      />
    </FormField>
  );
};
