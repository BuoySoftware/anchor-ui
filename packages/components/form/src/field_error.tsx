import compact from "lodash/compact";
import flatten from "lodash/flatten";
import { FieldError as HookFormFieldError } from "react-hook-form";
import snakeCase from "lodash/snakeCase";
import { BoxProps } from "@buoysoftware/anchor-layout";
import { TFunction } from "i18next";
import { useTranslation } from "react-i18next";

import { useFormScope } from "./form_scope_provider";
import { ErrorMessage } from "./error_message";

export type InputType =
  | "checkbox"
  | "collectionCheckbox"
  | "hidden"
  | "phoneNumber"
  | "radio"
  | "select"
  | "text";

interface OwnProps {
  error?: HookFormFieldError;
  name?: string;
  inputId: string;
  inputType?: InputType;
}

export type FieldErrorProps = OwnProps & BoxProps;

/**
 * Look up an error message in the i18next translations. This utilizes the
 * i18next fallback mechanism for determining who to translate the error message.
 *
 * The order is as follows:
 *
 * 1. An error message in a custom provided namespace via tNamespace scoped to
 *    forms.errors and then the form scope, field name, and error typea.
 *
 * 2. An error message in the default forms namespace scoped to the form scope,
 *    field name and error type.
 *
 * 3. An error message in default forms namespace for the given input type and
 *    error type.
 *
 * 4. An error message in the anchorForms namespace for the given input type and
 *    error type.
 *
 * 5. An error message in the forms namespace for the given error type
 *
 * 6. An error message in the anchorForms namespace for the given error type
 */
function lookupErrorMessage(
  error: HookFormFieldError,
  scope: string,
  name: string,
  inputType: InputType,
  t: TFunction,
  tNamespace: string[]
): string {
  if (error.message !== undefined && error.message !== "") {
    return error.message;
  }

  const i18nName = snakeCase(name);
  const i18nErrorScope = snakeCase(scope);

  // These are translations scoped to the form's scope and field name.
  // The use of forms.errors is to look into the custom tNamespace if one was
  // provided. The second will look up in the forms default namespace.
  const fieldScopes = [
    `forms.errors.${i18nErrorScope}.${i18nName}.${error.type}`,
    `errors.${i18nErrorScope}.${i18nName}.${error.type}`,
  ];

  // Global scopes are used to look up defaults in the forms namespace or
  // the anchorForms namespace
  const globalScopes = [
    `errors.input_types.${snakeCase(inputType)}.${error.type}`,
    `errors.${error.type}`,
  ];

  const combinedGlobalScopes = tNamespace.flatMap((namespace: string) =>
    globalScopes.map((scope) => `${namespace}:${scope}`)
  );

  return t([...fieldScopes, ...combinedGlobalScopes]);
}

export const FieldError: React.FC<FieldErrorProps> = ({
  error,
  inputId,
  name,
  inputType = "text",
  ...elementProps
}): React.ReactElement | null => {
  const { scope, tNamespace = [] } = useFormScope();
  const namespace = compact(flatten([tNamespace, "forms", "anchorForms"]));
  const { t } = useTranslation(namespace);

  if (!error || !name) return null;

  const message = lookupErrorMessage(
    error,
    scope,
    name,
    inputType,
    t,
    namespace
  );

  return (
    <ErrorMessage
      errors={message.split("\n")}
      parentId={inputId}
      {...elementProps}
    />
  );
};
