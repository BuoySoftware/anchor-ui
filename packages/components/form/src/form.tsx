import kebabCase from "lodash/kebabCase";
import { useEffect, useState } from "react";
import {
  FieldValues,
  FormProvider,
  get,
  SubmitHandler,
  UseFormReturn,
} from "react-hook-form";
import { ErrorMessage } from "./error_message";
import { Box, BoxProps } from "@buoysoftware/anchor-layout";

import { FormScopeProvider, TNamespace } from "./form_scope_provider";

export type FormValidationErrors<T> = {
  [key in keyof T]: string[];
};

export type ErrorKeys<TFormValues> = Record<string, keyof TFormValues>;

interface OwnProps<TFormValues extends FieldValues> {
  "data-testid"?: string;
  id: string;
  scope: string;
  onSubmit?: SubmitHandler<TFormValues>;
  children: React.ReactNode;
  form: UseFormReturn<TFormValues>;
  errorKeys?: ErrorKeys<TFormValues>;
  mutationErrors?: Partial<FormValidationErrors<TFormValues>>;
  tNamespace?: TNamespace;
}

export type FormProps<TFormValues extends FieldValues> = OwnProps<TFormValues> &
  BoxProps;

export const Form = <TFormValues extends FieldValues>({
  "data-testid": testId,
  id,
  scope,
  children,
  errorKeys = {},
  form: { handleSubmit, ...form },
  onSubmit = () => undefined,
  mutationErrors = {},
  tNamespace,
  ...elementProps
}: FormProps<TFormValues>): React.ReactElement => {
  const [hasErrors, setHasErrors] = useState(false);

  useEffect(() => {
    if (hasErrors) {
      Object.entries(mutationErrors).forEach(([key, errors]) => {
        if (key !== "base") {
          const messages = errors?.join("\n");
          const inputKey = get(errorKeys, key, key);

          form.setError(inputKey as any, {
            type: "server",
            message: messages,
          });
        }
      });
    }
  }, [hasErrors, mutationErrors]);

  useEffect(() => {
    setHasErrors(!!Object.keys(mutationErrors).length);
  }, [mutationErrors]);

  const baseErrors = mutationErrors["base"] || [];

  return (
    <FormScopeProvider value={{ formId: id, baseErrors, scope, tNamespace }}>
      <FormProvider {...form} handleSubmit={handleSubmit}>
        <Box
          as="form"
          autoComplete="off"
          id={id}
          data-testid={testId || kebabCase(scope)}
          onSubmit={handleSubmit(onSubmit)}
          {...elementProps}
        >
          {children}
        </Box>
      </FormProvider>
    </FormScopeProvider>
  );
};
