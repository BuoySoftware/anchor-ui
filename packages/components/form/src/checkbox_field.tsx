import { Box, BoxProps } from "@buoysoftware/anchor-layout";
import { RegisterOptions, useFormContext } from "react-hook-form";

import { useFormField } from "./use_form_field";
import { InlineLabel } from "./inline_label";
import { Checkbox } from "./checkbox";
import { FieldError } from "./field_error";

interface OwnProps {
  name: string;
  value?: string;
  rules?: RegisterOptions;
}

type CheckboxFieldProps = OwnProps & BoxProps;

export type CheckboxFieldValue = "1" | false;

export const DEFAULT_CHECKED_VALUE: CheckboxFieldValue = "1";

export const CheckboxField: React.FC<CheckboxFieldProps> = ({
  name,
  value = DEFAULT_CHECKED_VALUE,
  rules,
  ...boxProps
}): React.ReactElement => {
  const { label, error, inputId } = useFormField({ name });
  const { register } = useFormContext();

  return (
    <Box
      as="fieldset"
      data-testid={`checkbox-field-${inputId}`}
      display="flex"
      flexDirection="column"
      mb="form.fieldGap"
      {...boxProps}
    >
      <InlineLabel
        data-testid={`label-${inputId}`}
        display="flex"
        alignItems="center"
        htmlFor={inputId}
      >
        <Checkbox
          aria-describedby={error ? `error-${name}` : undefined}
          aria-invalid={error ? "true" : "false"}
          data-testid={inputId}
          id={inputId}
          mr="xxs"
          value={value}
          {...register(name, { shouldUnregister: true, ...rules })}
        />
        {label}
      </InlineLabel>
      <FieldError
        error={error}
        inputId={inputId}
        inputType="checkbox"
        mt="s"
        name={name}
      />
    </Box>
  );
};
