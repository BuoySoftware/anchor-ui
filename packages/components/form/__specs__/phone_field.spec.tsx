import i18n from "i18next";
import { initReactI18next } from "react-i18next";
import { useForm } from "react-hook-form";

import { PhoneField } from "../src/phone_field";
import { Form } from "../src/form";
import { act, render, waitFor, UserRenderResult } from "../test_utils";

describe("<PhoneField />", () => {
  const translations = {
    errors: {
      invalidPhoneNumber: "This phone number is invalid",
    },
    hints: {
      test_form: {
        phone_number: "Enter a 10 digit phone number",
      },
    },
    labels: {
      phone_number: "Cell Phone Number",
    },
    placeholders: {
      test_form: {
        phone_number: "ex: xxx-xxx-xxxx",
      },
    },
  };

  i18n.use(initReactI18next).init({
    lng: "en",
    ns: ["forms"],
    resources: {
      en: {
        forms: translations,
      },
    },
  });

  const MockForm: React.FC = (): React.ReactElement => {
    const form = useForm();

    return (
      <Form
        data-testid="test-form"
        form={form}
        id="test-form"
        onSubmit={jest.fn()}
        scope="test_form"
      >
        <PhoneField name="phoneNumber" />
        <button type="submit">Submit</button>
      </Form>
    );
  };

  function setup(): UserRenderResult {
    return render(<MockForm />);
  }

  const FIELD_LABEL = i18n.t("forms:labels.phone_number");

  it("renders properly", () => {
    const form = setup();

    expect(form.asFragment()).toMatchSnapshot();
  });

  context("with an error", () => {
    it("displays errors associated to the field", async () => {
      const { user, getByLabelText, getByText } = setup();
      const errorMessage = i18n.t("forms:errors.invalidPhoneNumber");

      await act(() => {
        user.click(getByText("Submit"));
      });

      await waitFor(() => {
        expect(getByLabelText(FIELD_LABEL)).toHaveErrorMessage(errorMessage);
      });
    });
  });

  context("with a hint", () => {
    it("displays a hint associated to the field", async () => {
      const { user, getByText } = setup();
      const hint = i18n.t("forms:hints.test_form.phone_number");

      await act(() => {
        user.click(getByText("Submit"));
      });

      await waitFor(() => {
        expect(getByText(hint)).toBeInTheDocument();
      });
    });
  });
});
