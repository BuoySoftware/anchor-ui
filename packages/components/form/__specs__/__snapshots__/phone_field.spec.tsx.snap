// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhoneField /> renders properly 1`] = `
<DocumentFragment>
  <form
    autocomplete="off"
    class="sc-bcXHqe gKnvGb"
    data-testid="test-form"
    id="test-form"
  >
    <div
      class="sc-bcXHqe eCUswC sc-ftTHYK"
      data-testid="phone-field-phone_number"
      display="flex"
      width="input"
    >
      <label
        class="sc-bcXHqe sc-hLBbgP sc-jSUZER sc-iBYQkv jMrRtF jvXitA"
        color="text.primary"
        font-weight="600"
        for="phone_number"
      >
        Cell Phone Number
      </label>
      <div
        class="sc-bcXHqe bRutrG"
      >
        <div
          class="sc-bcXHqe sc-jrcTuL gKnvGb gAYoPi"
          data-testid="input-wrapper-phone-number"
        >
          <div
            class="sc-bcXHqe sc-gswNZR dxGagn hcNfAU"
            display="flex"
          >
            <div
              class="sc-bcXHqe crvKEL"
            >
              <input
                aria-invalid="false"
                class="sc-pyfCe jTfFQB"
                data-testid="input-phone-number"
                id="phone_number"
                maxlength="16"
                name="phoneNumber"
                placeholder="ex: xxx-xxx-xxxx"
                type="tel"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <p
        class="sc-bcXHqe sc-hLBbgP sc-eDvSVe gmdqHQ iwgFby"
        color="text.secondary"
        font-size="font-size-75"
        font-style="none"
        font-weight="400"
      >
        Enter a 10 digit phone number
      </p>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
